File: java/Solution973.java
package leetCode.java;

import java.util.*;

//Heap O(nlogk) O(k)
class Solution973 {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(
            (a, b) -> (b[0] * b[0] + b[1] * b[1]) - (a[0] * a[0] + a[1] * a[1]) 
        ); // max heap, 依照距離大到小
        for (int[] point : points) {
            pq.offer(point);
            if (pq.size() > k) { // 拿掉大的
                pq.poll();
            }
        }

        int[][] res = new int[pq.size()][2];
        for (int i = 0; i < res.length; ++i) {
            res[i] = pq.poll();
        }
        return res;
    }
}

File: c++/Solution973.cpp
//Heap O(nlogk) O(k)
class Solution973 {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        //{距離, points index}, 大排到小, pair會先比較first再second
        priority_queue<pair<int,int>, vector<pair<int,int>>, less<pair<int,int>>> pq;
        for(int i=0; i<points.size(); i++){
            int len = points[i][0] * points[i][0] + points[i][1] * points[i][1];
            pq.push({len, i});

            if(pq.size() > k)
                pq.pop();
        }

        vector<vector<int>> res;
        while(!pq.empty()){
            auto data = pq.top(); pq.pop();
            res.push_back(points[data.second]);
        }
        return res;
    }
};